### PR TITLE
Cache CloudStorage credentials

### PR DESCRIPTION
**Pull request by lavoiesl**
The credentials token source has built in caching, only generating a token when needed.

Caching the credentials object allows to reuse the same source for the same scopes, leveraging that caching.

Otherwise, CloudStorageFS was generating new tokens every time a method was being called, wasting a significant amount of CPU.

### CODE REVIEW

**Review by bmassemin — APPROVED** [review]
I'm wondering if it could be more simple and performant to give all the required scopes in `NewCloudStorageFS`, build a client, and then cache the client instead of the credentials (if I understand correctly the code).

**Comment by lavoiesl** [comment]
It's doable, but it would make the api a bit harder to use the constructor of the object would need to know ahead of time which scope it would require. It could be a valid, I’m not sure. That being said, it'd be a significantly larger change since it would break the constructor, so perhaps not do it in the scope of this PR.

**Comment by lavoiesl** [comment]
Yesterday, we ran a load test; I took a second look at our profiler.

For the top 50% weight (to reduce noise):

| Operation                | CPU       | Allocated Heap |
|--------------------------|-----------|----------------|
| Save                     | 335.57 ms | 39.2 MiB       |
| JSON encode              | 97.83 ms  | 17.87 MiB      |
| Create (without writing) | 175.22 ms | 17.91 MiB      |
| bucketHandle             | 150.26 ms | 14.6 MiB       |
| client                   | 137.04 ms | 12.38 MiB      |
| findCredentials          | 135.65 ms | 12.28 MiB      |
| ObjectsInsertCall.doRequest          | 3.83 s    | 126.5 MiB      |
| jwtSource.Token          | 3.59 s    | 126.5 MiB      |

Finding credentials takes about 77% of the "Create" CPU time:
image

Token generation takes up most of the upload CPU time:
image

Similar story for allocated heap:
image
image

---

**Conclusion:** Credentials recreation is the culprit and client recreation is negligible

File: pkg/cloudstorage_fs.go [hunk 1]
```diff
@@ -161,7 +179,7 @@ func (c *cloudStorageFS) client(ctx context.Context, scope string, extraScopes .
 		return nil, nil, errors.Wrap(err, "cloud storage: unable to retrieve default token source")
 	}
 
-	client, err := gcp.NewHTTPClient(gcp.DefaultTransport(), gcp.CredentialsTokenSource(creds))
+	client, err := gcp.NewHTTPClient(gcp.DefaultTransport(), creds.TokenSource)
```
Comment (yousifh) [inline pkg/cloudstorage_fs.go:182]: Was `gcp.CredentialsTokenSource(creds)` doing anything special? I think it's just returning a type wrapper of creds.TokenSrouce?

**Review by yousifh — APPROVED** [review]

**Review by pior — COMMENTED** [review]
The OAuth token provided by the metadata server are only valid for a short period of time (1h ?).

We can't cache them without a mechanism to evict or refresh them.

**Comment by lavoiesl** [comment]
The client is re-instantiated with a `TokenSource`, not a `Token`. It's the `TokenSource` that is responsible to cache/refresh tokens, not the client.

```go
// A TokenSource is anything that can return a token.
type TokenSource interface {
	// Token returns a token or an error.
	// Token must be safe for concurrent use by multiple goroutines.
	// The returned Token must not be modified.
	Token() (*Token, error)
}

// ReuseTokenSource returns a TokenSource which repeatedly returns the
// same token as long as it's valid, starting with t.
// When its cached token is invalid, a new token is obtained from src.
//
// ReuseTokenSource is typically used to reuse tokens from a cache
// (such as a file on disk) between runs of a program, rather than
// obtaining new tokens unnecessarily.
//
// The initial token t may be nil, in which case the TokenSource is
// wrapped in a caching version if it isn't one already. This also
// means it's always safe to wrap ReuseTokenSource around any other
// TokenSource without adverse effects.
func ReuseTokenSource(t *Token, src TokenSource) TokenSource {
	// Don't wrap a reuseTokenSource in itself. That would work,
	// but cause an unnecessary number of mutex operations.
	// Just build the equivalent one.
	if rt, ok := src.(*reuseTokenSource); ok {
		if t == nil {
			// Just use it directly.
			return rt
		}
		src = rt.new
	}
	return &reuseTokenSource{
		t:   t,
		new: src,
	}
}
```

**Comment by lavoiesl** [comment]
Tested this PR with a benchmark from my computer: https://github.com/Shopify/go-storage/pull/21

```
go test -bench=BenchmarkCloudStorageFS -benchtime=4s -cpuprofile=cpu.out -benchmem -memprofile=mem.out ./...

Before
BenchmarkCloudStorageFS/create-8         	      18	 268202072 ns/op	 8575851 B/op	     857 allocs/op
BenchmarkCloudStorageFS/read-8           	      38	 119780156 ns/op	  149152 B/op	     658 allocs/op

After
BenchmarkCloudStorageFS/create-8         	      24	 192845307 ns/op	 8434111 B/op	     400 allocs/op
BenchmarkCloudStorageFS/read-8           	      70	  67424345 ns/op	   19537 B/op	     211 allocs/op
```

Writes are 28% faster, reads are 44% faster.

Comparing the profiles with `go tool pprof -http 127.0.0.1:1234 -base cpu.before.out cpu.out`, the `Transport.RoundTrip` goes from about 17% of the CPU usage to essentially 0, so I think it demonstrates that while not being ideal, this PR is sufficient to fix the issue.

**Comment by pior** [comment]
> TokenSource

Right, LGTM then 👍

**Review by erikaleigh — APPROVED** [review]
TIL this repo exists! 🙃 

I mean, this looks okay to me, just catching up on the convo and the tests you performed. 👍

**Pull request closed without merging**